_drafts/2024-09-01-aws-terraform-02.md

```diff
@@ -0,0 +1,267 @@
+
+Now that we can connect to ECR, we need to provide access for our EC2 instance to run commands against ECR.
+
+In your `infra` directory create a new file `iam.tf` and add the following:
+
+```terraform
+# /infra/iam.tf
+
+resource "aws_iam_role" "ec2_ecr_readonly_role" {
+  name               = "ec2-ecr-role"
+  assume_role_policy = data.aws_iam_policy_document.ec2_assume_role_policy.json
+
+  inline_policy {
+    name   = "ecr-read-policy"
+    policy = data.aws_iam_policy_document.ec2_container_registry_read_only.json
+  }
+}
+
+data "aws_iam_policy_document" "ec2_assume_role_policy" {
+  statement {
+    actions = ["sts:AssumeRole"]
+
+    principals {
+      type        = "Service"
+      identifiers = ["ec2.amazonaws.com"]
+    }
+  }
+}
+
+data "aws_iam_policy_document" "ec2_container_registry_read_only" {
+  statement {
+    actions = [
+      "ecr:GetAuthorizationToken",
+      "ecr:BatchGetImage",
+      "ecr:GetDownloadUrlForLayer",
+    ]
+    resources = ["*"]
+  }
+}
+```
+
+What we have done is defined an [aws_iam_role](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/iam_role) and applied a couple of [aws_iam_policy_documents](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/iam_policy_document) which:
+
+1. Define which services can assume this role (`ec2_assume_role_policy`). In this case only EC2 services can assume the role
+2. Define the actions a service with role can perform (`ec2_container_registry_read_only`). In this case the service only requires one action to get an auth token for Docker to read the image.
+
+We've defined our role and appropriate policies for the role, now we need to apply it to our EC2 instance.
+
+Open `unicorn-api.tf` and add the following resource: 
+
+```terraform
+# /infra/unicorn-api.tf
+
+resource "aws_iam_instance_profile" "unicorn_api_iam_profile" {
+  name = "unicorn-api-iam-profile"
+  role = aws_iam_role.ec2_readonly_role.name
+}
+```
+
+
+Now add the [iam instance profile](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/iam_instance_profile) to the aws instance.
+
+```terraform
+# /infra/unicorn-api.tf
+
+resource "aws_instance" "unicorn_api" {
+
+  # ...
+
+  iam_instance_profile = aws_iam_instance_profile.unicorn_api_iam_profile.name
+ 
+  # ...
+}
+```
+
+Great! Now that we've defined the IAM role and applied it to our EC2 resource, we can define a script for our instance to run on start up which will pull the ECR image and run it in docker.
+
+Create a new file in the `infra` directory, `init-unicorn-api.sh`
+
+```bash
+#!/bin/bash
+
+aws ecr get-login-password --region [region] | docker login --username AWS --password-stdin [ecr repository url]
+docker run -p 80:8080 [ecr repository url]:latest
+```
+
+Now add the following permissions for the script:
+
+`chmod 755 init-unicorn-api.sh`
```

Comment:
Do I actually need to set the file permissions and if so is this the right permission?